apifest-oauth20/src/main/java/com/apifest/oauth20/persistence/cassandra/CassandraDBManager.java

```diff
@@ -0,0 +1,510 @@
+            if(iter.hasNext()) {
+                AccessToken atoken = new AccessToken();
+                Row row = iter.next();
+                mapRowToAccessToken(row, atoken);
+                return atoken;
+            }
+        } catch(Throwable e) {
+            log.error(e.getMessage(), e);
+            return null;
+        }
+        return null;
+    }
+    private void mapRowToAccessToken(Row row, AccessToken atoken) {
+        atoken.setToken(row.getString("access_token"));
+        atoken.setRefreshToken(row.getString("refresh_token"));
+        atoken.setExpiresIn(row.getString("expires_in"));
+        atoken.setScope(row.getString("scope"));
+        atoken.setType(row.getString("type"));
+        atoken.setValid(row.getBool("valid"));
+        atoken.setClientId(row.getString("client_id"));
+        atoken.setCodeId(row.getString("code_id"));
+        atoken.setUserId(row.getString("user_id"));
+        atoken.setDetails(row.getMap("details", String.class, String.class));
+        atoken.setCreated(row.getTimestamp("created").getTime());
+        atoken.setRefreshExpiresIn(row.getString("refresh_expires_in"));
+    }
+
+    @Override
+    public void removeAccessToken(String accessToken) {
+        try {
+            Delete.Where stmt = QueryBuilder.delete().from(KEYSPACE_NAME, ACCESS_TOKEN_TABLE_NAME)
+                    .where(QueryBuilder.eq("access_token", accessToken));
+            session.execute(stmt);
+        } catch(Throwable e) {
+            log.error(e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public AccessToken findAccessTokenByRefreshToken(String refreshToken, String clientId) {
+        try {
+            Select.Where stmt = QueryBuilder.select().from(KEYSPACE_NAME, ACCESS_TOKEN_TABLE_NAME)
+                    .allowFiltering() // TODO: build a materialized view with refreshToken + clientId key
+                    .where()
+                        .and(QueryBuilder.eq("refresh_token", refreshToken))
+                        .and(QueryBuilder.eq("client_id", clientId))
+            ;
+            ResultSet rs = session.execute(stmt);
+            Iterator<Row> iter = rs.iterator();
+            if(iter.hasNext()) {
+                AccessToken atoken = new AccessToken();
+                Row row = iter.next();
+                mapRowToAccessToken(row, atoken);
+                log.debug(atoken.getToken());
+                return atoken;
+            }
+        } catch(Throwable e) {
+            log.error(e.getMessage(), e);
+            return null;
+        }
+        return null;
+    }
+
+    @Override
+    public List<AccessToken> getAccessTokenByUserIdAndClientApp(String userId, String clientId) {
+        // TODO: build a materialized view with userId + clientId key
+        List<AccessToken> list = new ArrayList<AccessToken>();
+        try {
+            Select.Where stmt = QueryBuilder.select().from(KEYSPACE_NAME, ACCESS_TOKEN_TABLE_NAME)
```

Comment:
Considering this needs to be very fast, because it's executed on every request, why don;t you consider a composite key that includes the token, user_id and client id. If you put the client_id first, you'd be able to get all the tokens for a client without using a secondary index AND fetch a specific token by filling out the rest of the row key.